// ngx-alerts/esm2015/lib/service/alert.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { BehaviorSubject, Subject, timer } from 'rxjs';
import { ALERT_CONFIG } from '../alert.config';
import { scan, take } from 'rxjs/internal/operators';
import { AlertReducer } from './alert.reducer';
export class AlertService {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.dispatcher = new Subject();
        this.state = new BehaviorSubject([]);
        this.initConfig();
        this.dispatcher
            .pipe(scan(AlertReducer.reduce, []))
            .subscribe(this.state);
    }
    /**
     * @private
     * @return {?}
     */
    initConfig() {
        if (!this.config) {
            this.config = {};
        }
        this.config.timeout = !!this.config.timeout ? this.config.timeout : 5000;
        this.config.maxMessages = !!this.config.maxMessages ? this.config.maxMessages : 5;
    }
    /**
     * @return {?}
     */
    get messages() {
        return this.state;
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    info(msg) {
        this.addAlert({ content: msg, type: 'info' });
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    danger(msg) {
        this.addAlert({ content: msg, type: 'danger' });
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    success(msg) {
        this.addAlert({ content: msg, type: 'success' });
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    warning(msg) {
        this.addAlert({ content: msg, type: 'warning' });
    }
    /**
     * @param {?} alert
     * @return {?}
     */
    close(alert) {
        this.dispatcher.next({ fn: AlertReducer.remove, alert: alert, config: this.config });
    }
    /**
     * @private
     * @param {?} alert
     * @return {?}
     */
    addAlert(alert) {
        this.dispatcher.next({ fn: AlertReducer.add, alert: alert, config: this.config });
        timer(this.config.timeout)
            .pipe(take(1))
            .subscribe(() => {
            this.dispatcher.next({ fn: AlertReducer.remove, alert: alert, config: this.config });
        });
    }
}
AlertService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AlertService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [ALERT_CONFIG,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AlertService.prototype.dispatcher;
    /**
     * @type {?}
     * @private
     */
    AlertService.prototype.state;
    /**
     * @type {?}
     * @private
     */
    AlertService.prototype.config;
}
